src/tail/tunnel.rs

```diff
@@ -0,0 +1,73 @@
+use std::path::PathBuf;
+use std::process::Stdio;
+use std::str;
+
+use log::log_enabled;
+use log::Level::Info;
+use tokio::process::Child;
+use tokio::process::Command;
+use tokio::sync::oneshot::Receiver;
+
+pub struct Tunnel {
+    child: Child,
+}
+
+impl Tunnel {
+    pub fn new() -> Result<Tunnel, failure::Error> {
+        let tool_name = PathBuf::from("cloudflared");
+        // TODO: Finally get cloudflared release binaries distributed on GitHub so we could
+        // simply uncomment the line below.
+        // let binary_path = install::install(tool_name, "cloudflare")?.binary(tool_name)?;
+
+        // TODO: allow user to pass in their own ports in case ports 8080 (used by cloudflared)
+        // and 8081 (used by cloudflared metrics) are both already being used.
+        let args = ["tunnel", "--metrics", "localhost:8081"];
+
+        let mut command = command(&args, &tool_name);
+        let command_name = format!("{:?}", command);
+
+        let child = command
+            .spawn()
+            .expect(&format!("{} failed to spawn", command_name));
+
+        Ok(Tunnel { child })
+    }
+
+    pub async fn run(self, rx: Receiver<()>) -> Result<(), failure::Error> {
+        rx.await?;
+        self.shutdown().await
+    }
+
+    pub async fn shutdown(mut self) -> Result<(), failure::Error> {
+        // eprintln!("killing cloudflared");
+        if let Err(e) = self.child.kill() {
```

Comment:
TODO: consider a small delay on this kill; since cloudflared is a child process it receives the signal to shut down independently of its parent process (wrangler) but its "graceful" shutdown is relatively long and we do this double tap because it is,  as of now, relatively harmless; we still want to wait on it to exit though hence  the wait_with_output after.